README.md

```diff
@@ -15,17 +15,40 @@ Implement the tool `st` that does exact pattern matching using a suffix tree. Te
 
 ## Specify if you have used a linear time or quadratic time algorithm.
 
+Quadratic time naive insert algorithm. McCreight's algorithm would have been cool, but time, unfortunately, is limited.
+
 ## Insights you may have had while implementing and comparing the algorithms.
 
+N/A
+
 ## Problems encountered if any.
 
+Testing of trees is quite difficult. Initially, a test was made in order to capture the "semantics" of a suffix tree. That is; we do not care about which order nodes
+are in. We only care about who their parent and children are (and their edge labels). This lead to a quite complicated test, which utilized prime factors in order
+to reliably decouple the test from the implementation. The reasoning is, that the test was supposed to be written before the implementation - and would aid in actually
```

Comment:
I fail to see how number theory plays into suffix trees, but it is quite simple to test a tree without making assumptions about the order of children. A tree should have the numbers 0, 1, ..., n (or n+1, depending on how you count) as leaves, and the path to each leaf should match the corresponding suffix. A tree should have at most 2n-1 nodes. For each inner node, the path to that node should match a prefix of the suffixes in the leaves under it. There are lots of tests like these.

Reply:
Right, but I felt like that being a catch-22. In order to assert if a suffix is correctly inserted in the tree (which requires searching), we would make assertions where we search (correctly, that is) for edges and travel down the tree.

So instead, maybe unwisely, I ended up with asserting that after each insert, that the tree looks correctly.
In order to do so, and where number theory suddenly comes into play, I can show an example below:

Say we have a tree with root _r_ and three edges _a, b, c_ to the root's three children.
If we assert(edge1, a), assert(edge2, b) and assert(edge3, c), then it is very implementation dependent, that the tree looks like that after some _k_ suffix insertions. Instead I assigned for instance a prime number for each edge; a=2, b=3, c=5. We then check if (edge1, a)*(edge2,b)*(edge3,c)=30. This however only tells us they exist. To retrieve the correct node we then have to map the above to some expected locations... if that makes any sense. The input could have been (a,c,b) but we want it in some expected order (a,b,c) to test. 

Of course this was way to complicated, which I realized soon - thus I thought that proper decoupled testing of trees is hard. Maybe I should just have implemented a searching algorithm for the tests :)

Reply:
I fail to see the sanity in assigning primes to edges. The edges are uniquely defined already by their labels. They are also ordered if you want that. They might not be ordered in the linked lists, but they have an order nevertheless, and you can determine it by the first character.

If you want a test to state that a node $v$ has three children, $w_1, w_2, w_3$ it is trivial to test that. Equally trivial to check if they have the right edge labels. Trying to abstract away from the actual structure when the structure is precisely what you want to implement doesn't seem like a fruitful endeavour.

You have a tendency to over-abstract and over-engineer things. I recognise it because I suffer from the same character defect. I have found that in 99% of cases, it takes ten times longer to do it that way compared to implementing the most concrete solution you can think of. If abstraction is needed later, you can spend the time on it there. Up-front abstraction practically never pays off. At least, it never has for me, and I don't know anyone who honestly claims that it has for them.

You don't need much of an implementation of a tree to be able to traverse it and get hold of all leaves. That is something anyone can code up in 10 minutes. So that is not worth abstracting anything over. Once you can get all the leaves, you can check if their labels are 0, 1, 2, ..., n using a bit vector, a set, or whatever you have at hand. That is another 10 minutes. If it takes more than 20 minutes to write this test, then you are doing it wrong.

To check if the path labels are correct, you need a little more work. In Python it is ~5 lines, but in C it will be a bit more. You just run up from the leaves (and if you haven't added parent pointers yet, you spend ten minutes on adding them now). Each node has one edge going up to the parent, you extract the range there, and put it in an array. Then you allocate a buffer for the path label, run through the ranges you collected in reverse order, copy from the original string the range they specify into the new buffer, and when you are done you have the path label. There are some C issues with allocation and such, but it won't take 30 minutes to implement that test either.

If you can extract path labels from leaves you can also do it from any other node, unless you did something crazy. So checking if the path labels at inner nodes are prefixes of the leaves is something you can do 30 minutes later.

If you keep the thing concrete, you can write tests that verify practically everything about a suffix tree in a couple of hours. You don't need to have the construction algorithm done before writing these tests; you just need to settle on what the variable that refers to the range and the parent is called and how you represent the children.

Overthinking tasks like these will take forever, and in the end, you don't have any working implementation. It was the same with the profile HMMs; 99% of the project was overthinking something that could be implemented in an afternoon by having a table and a scan along the genome.

Your life will get easier if you start implementing things in the simplest way possible. It is easy to refactor running code, but until the code is running, even the most beautiful mental models are useless.

Reply:
I have much to learn 😁

In regards to profile HMM's it was hard from the get-go due to language choices. I at least thought it was neat, that it was solved in a functional paradigm, as correctness could be shown for some functions. However, in reality, it isn't that vital - I know. Now I can write in C, so that (or C++) would be my language of choice to such tasks.

I'll take note and try to avoid over-engineering things. It is sometimes, admittedly, a fun part to over-abstract.

Reply:
It is heaps of fun. Just rarely productive :)